alibi_detect/cd/base.py

```diff
@@ -86,13 +97,23 @@ def __init__(
         if n_folds is not None and n_folds > 1 and not retrain_from_scratch:
             raise ValueError("If using multiple folds the model must be retrained from scratch for each fold.")
 
-        # optionally already preprocess reference data
-        self.p_val = p_val
-        if preprocess_x_ref and isinstance(preprocess_fn, Callable):  # type: ignore[arg-type]
+        # x_ref preprocessing logic
+        self.preprocess_at_pred = not preprocess_at_init and not x_ref_preprocessed and preprocess_fn is not None
+        self.preprocess_at_init = preprocess_at_init and not x_ref_preprocessed and preprocess_fn is not None
+        self.x_ref_preprocessed = x_ref_preprocessed
+        # Check if preprocess_fn is valid
+        if (self.preprocess_at_init or self.preprocess_at_pred) \
+                and not isinstance(preprocess_fn, Callable):  # type: ignore[arg-type]
+            raise ValueError("`preprocess_fn` is not a valid Callable.")
+
+        # optionally preprocess reference data (now instead of at predict)
+        if self.preprocess_at_init:
```

Comment:
Yes it pretty much is. That's a nice idea, I'll have a look at doing that. 

I initially wanted to add a specific drift detector base class to avoid duplication, but decided this would be too big a change for the PR. A mix in would be a nice solution I think...

Reply:
Done in https://github.com/SeldonIO/alibi-detect/pull/458

Reply:
I've added a mixin for the duplication in the `get_config` methods, based on your comment below. 

However, for the duplication in this case, I think this is highlighting the fact that it might make sense for us to have a base drift detector class. Similar duplication was present before this PR, so personally, I'd be in favor of leaving this out of this PR.

Reply:
Discussed in #458